memory/ndctl_selftest.py

```diff
@@ -0,0 +1,121 @@
+#!/usr/bin/env python
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: 2019 IBM
+# Author: Harish <harish@linux.vnet.ibm.com>
+#
+#
+
+
+import os
+import json
+
+from avocado import Test
+from avocado import main
+from avocado.utils import process, build, distro, git
+from avocado.utils.software_manager import SoftwareManager
+
+
+class NdctlTest(Test):
+    """
+    Ndctl is a user space tool to manage persistent memory devices
+    This test uses the selftests of the userspace tool to sanity check the
+    persistent memory
+    """
+
+    def get_bus_ids(self):
+        """
+        Get the provider IDs for each pmem device
+        """
+        bus_ids = []
+        os.chdir(self.sourcedir)
+        json_op = json.loads(process.system_output(
+            './ndctl/ndctl list -B', shell=True))
+        for nid in json_op:
+            for key, value in nid.items():
+                if key == 'provider':
+                    bus_ids.append(value)
+        return bus_ids
+
+    def setUp(self):
+        """
+        Prequisite for ndctl selftest on non-NFIT devices
+        """
+        smg = SoftwareManager()
+        self.url = self.params.get(
+            'url', default="https://github.com/pmem/ndctl.git")
+        self.branch = self.params.get('branch', default='master')
+
+        deps = ['gcc', 'make', 'automake', 'autoconf', 'patch']
+        detected_distro = distro.detect()
+
+        if 'SuSE' in detected_distro.name:
+            deps.extend(['ruby2.5-rubygem-asciidoctor', 'libtool',
+                         'libkmod-devel', 'libudev-devel', 'keyutils-devel',
+                         'libuuid-devel-static', 'libjson-c-devel',
+                         'systemd-devel', 'kmod-bash-completion'])
+        else:
+            # TODO: Add RHEL when support arrives
+            self.cancel('Unsupported OS %s' % detected_distro.name)
+
+        for package in deps:
+            if not smg.check_installed(package) and not smg.install(package):
+                self.cancel(
+                    "Fail to install %s required for this test." % (package))
```

Comment:
can we start using python3 based string formate can be like 

`self.cancel("Fail to install {} required for this test." .format(package))`

Reply:
I think `.format` is supported from python2 time frame as well, not sure how it is related to python3. Anyhow if you are suggesting this as string format specifically it would be good to handle all other files with this change.

Reply:
yes it supported from python2 but python3 onwards it recomended to use

Reply:
It would be helpful if you could point me to the recommendation link so that I could send a patch addressing other tests as well. Thanks.

Reply:
you can refer this `https://www.python.org/dev/peps/pep-0498/`

Reply:
https://realpython.com/python-string-formatting/